File: src/main/java/com/anod/appwatcher/backup/GDriveSync.java
package com.anod.appwatcher.backup;

import android.app.Activity;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.support.v4.app.NotificationCompat;
import android.widget.Toast;

import com.anod.appwatcher.R;
import com.anod.appwatcher.backup.gdrive.SyncConnectedWorker;
import com.anod.appwatcher.backup.gdrive.SyncTask;
import com.anod.appwatcher.utils.GooglePlayServices;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.drive.Drive;

/**
 * @author alex
 * @date 1/19/14
 */
public class GDriveSync extends GooglePlayServices implements SyncTask.Listener {


    private static final int ACTION_SYNC = 2;
    private static final int NOTIFICATION_ID = 2;

    private Listener mListener;

    public void showResolutionNotification(PendingIntent resolution) {
        NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext);
        builder
                .setAutoCancel(true)
                .setSmallIcon(R.drawable.ic_stat_update)
                .setContentTitle("Google Drive sync failed.")
                .setContentText("Required user action")
                .setContentIntent(resolution)
        ;

        Notification notification = builder.build();
        NotificationManager mNotificationManager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
        mNotificationManager.notify(NOTIFICATION_ID, notification);
    }


    public interface Listener {
        void onGDriveConnect();
        void onGDriveSyncProgress();
        void onGDriveSyncStart();
        void onGDriveSyncFinish();
		void onGDriveError();
	}

    public GDriveSync(Activity activity, Listener listener) {
        super(activity);
        mListener = listener;
    }

    public GDriveSync(Context context) {
        super(context);
    }

    @Override
    protected void onConnectAction(int action) {
        mListener.onGDriveConnect();
        if (action == ACTION_SYNC) {
            new SyncTask(mContext, this, createGoogleApiClientBuilder().build()).execute(true);
        }
    }

    @Override
    protected void onConnectionError() {
        mListener.onGDriveError();
    }

    public void sync() {
        mListener.onGDriveSyncStart();
        if (!isConnected()) {
            connectWithAction(ACTION_SYNC);
        } else {
            new SyncTask(mContext, this, createGoogleApiClientBuilder().build()).execute(true);
        }
    }

    public void syncLocked() throws Exception {
        if (!isConnected()) {
            connectLocked();
        }
        SyncConnectedWorker worker = new SyncConnectedWorker(mContext, mGoogleApiClient);
        worker.doSyncInBackground();
    }

    protected GoogleApiClient.Builder createGoogleApiClientBuilder() {
        return new GoogleApiClient.Builder(mContext)
                .addApi(Drive.API)
                .addScope(Drive.SCOPE_APPFOLDER);
    }

    @Override
    public void onResult(SyncTask.Result result) {
        if (result == null) {
            //Connection error
            mListener.onGDriveError();
            return;
        }
        if (result.status) {
            mListener.onGDriveSyncFinish();
        } else {
            Toast.makeText(mContext, result.ex.getMessage(), Toast.LENGTH_SHORT).show();
            mListener.onGDriveError();
        }
    }

}

File: src/main/java/com/anod/appwatcher/backup/GDriveSync.kt
package com.anod.appwatcher.backup

import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.support.v4.app.NotificationCompat
import android.widget.Toast
import com.anod.appwatcher.R
import com.anod.appwatcher.backup.gdrive.*
import com.anod.appwatcher.utils.GooglePlayServices
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.drive.Drive

/**
 * @author alex
 * *
 * @date 1/19/14
 */
class GDriveSync(private val context: Context): SyncTask.Listener, UploadTask.Listener, GooglePlayServices.Client {

    var listener: Listener? = null
    val playServices = GooglePlayServices(context, this)

    val isSupported: Boolean
        get() = playServices.isSupported

    val playServiceStatusText: String
        get() = playServices.errorCodeText

    fun showResolutionNotification(resolution: PendingIntent) {
        val builder = NotificationCompat.Builder(context, NotificationChannel.DEFAULT_CHANNEL_ID)
        builder
                .setAutoCancel(true)
                .setSmallIcon(R.drawable.ic_notification)
                .setContentTitle(context.getString(R.string.google_drive_sync_failed))
                .setContentText(context.getString(R.string.user_action_required))
                .setContentIntent(resolution)

        val notification = builder.build()
        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.notify(NOTIFICATION_ID, notification)
    }

    interface Listener {
        fun onGDriveConnect()
        fun onGDriveSyncProgress()
        fun onGDriveSyncStart()
        fun onGDriveSyncFinish()
        fun onGDriveError()
    }

    fun connect() {
        playServices.connect()
    }

    fun sync() {
        listener?.onGDriveSyncStart()
        if (!playServices.isConnected) {
            playServices.connectWithAction(ACTION_SYNC)
        } else {
            SyncTask(context, this, createGoogleApiClientBuilder().build()).execute()
        }
    }

    fun upload() {
        listener?.onGDriveSyncStart()
        if (!playServices.isConnected) {
            playServices.connectWithAction(ACTION_UPLOAD)
        } else {
            UploadTask(context, this, createGoogleApiClientBuilder().build()).execute()
        }
    }

    @Throws(Exception::class)
    fun syncLocked() {
        if (!playServices.isConnected) {
            playServices.connectLocked()
        }
        val worker = SyncConnectedWorker(context, playServices.googleApiClient)
        worker.doSyncInBackground()
    }

    @Throws(Exception::class)
    fun uploadLocked() {
        if (!playServices.isConnected) {
            playServices.connectLocked()
        }
        val worker = UploadConnectedWorker(context, playServices.googleApiClient)
        worker.doUploadInBackground()
    }

    override fun createGoogleApiClientBuilder(): GoogleApiClient.Builder {
        return GoogleApiClient.Builder(context)
                .addApi(Drive.API)
                .addScope(Drive.SCOPE_APPFOLDER)
    }

    override fun onConnectAction(action: Int) {
        listener?.onGDriveConnect()
        if (action == ACTION_SYNC) {
            SyncTask(context, this, createGoogleApiClientBuilder().build()).execute()
        }
    }

    override fun onConnectionError() {
        listener?.onGDriveError()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        playServices.onActivityResult(requestCode, resultCode, data)
    }

    override fun onSyncTaskResult(result: ApiClientAsyncTask.Result) {
        if (result.status) {
            listener?.onGDriveSyncFinish()
        } else {
            Toast.makeText(context, result.ex?.message ?: "Error", Toast.LENGTH_SHORT).show()
            listener?.onGDriveError()
        }
    }

    override fun onUploadTaskResult(result: ApiClientAsyncTask.Result) {
        if (result.status) {
            listener?.onGDriveSyncFinish()
        } else {
            Toast.makeText(context, result.ex?.message ?: "Error", Toast.LENGTH_SHORT).show()
            listener?.onGDriveError()
        }
    }

    companion object {
        private const val ACTION_SYNC = 2
        private const val ACTION_UPLOAD = 3
        private const val NOTIFICATION_ID = 2
    }
}